gpbft/validator.go

```diff
@@ -0,0 +1,294 @@
+package gpbft
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math"
+	"sync/atomic"
+
+	"github.com/filecoin-project/go-f3/internal/caching"
+	"go.opentelemetry.io/otel/metric"
+)
+
+var (
+	_ MessageValidator = (*cachingValidator)(nil)
+	_ ProgressObserver = (*cachingValidator)(nil)
+)
+
+// ProgressObserver defines an interface for observing and being notified about
+// the progress of a GPBFT instance as it advances through different instance,
+// rounds or phases.
+type ProgressObserver interface {
+	// NotifyProgress is called to notify the observer about the progress of GPBFT
+	// instance, round or phase.
+	NotifyProgress(instance, round uint64, phase Phase)
+}
+
+type cachingValidator struct {
+	progress atomic.Pointer[progress]
+	// cache is a bounded cache of messages that have already been validated, grouped
+	// by instance.
+	cache             *caching.GroupedSet
+	committeeLookback uint64
+	committeeProvider CommitteeProvider
+	networkName       NetworkName
+	signing           Signatures
+}
+
+type progress struct {
+	id    uint64
+	round uint64
+	phase Phase
+}
+
+func newValidator(host Host, cp CommitteeProvider, committeeLookback uint64, maxCachedInstances int, maxCachedMsgsPerInstance int) *cachingValidator {
+	validator := cachingValidator{
+		cache:             caching.NewGroupedSet(maxCachedInstances, maxCachedMsgsPerInstance),
+		committeeProvider: cp,
+		committeeLookback: committeeLookback,
+		networkName:       host.NetworkName(),
+		signing:           host,
+	}
+	// Default to instance 0, round 0 and INITIAL phase.
+	validator.progress.Store(&progress{})
+	return &validator
+}
+
+func (v *cachingValidator) NotifyProgress(instance uint64, round uint64, phase Phase) {
+	v.progress.Store(&progress{
+		id:    instance,
+		round: round,
+		phase: phase,
+	})
+	if instance > 0 {
+		// Remove cache of validated messages for instance that are older than the
+		// previous instance.
+		v.cache.RemoveGroupsLessThan(instance - 1)
+	}
+}
+
+// ValidateMessage checks if the given message is valid. If invalid, an error is
+// returned. ErrValidationInvalid indicates that the message will never be valid
+// invalid and may be safely dropped.
+func (v *cachingValidator) ValidateMessage(msg *GMessage) (valid ValidatedMessage, err error) {
```

Comment:
We should still catch panics somewhere, although that can probably happen in the top-level runner in `validatePubsubMessage`.

Reply:
Looks like i fat fingered the logic that existed there; restored now. Apologies.